### DISCUSSION TITLE
[5.x]: Make it easier to compose a custom element chip or label

### DISCUSSION THREAD

**Discussion by engram-design** [craftcms/cms#14091] [Ideas]
With Craft 5, there's a concept of element chips, but it's still a bit tricky (or a lot of work) to compose a custom one, or tweak the existing one without re-writing everything. Most of the useful [Cp](https://github.com/craftcms/cms/blob/5.0/src/helpers/Cp.php) methods are private, like `baseElementAttributes()`, `elementStatusHtml()`, `elementLabelHtml()`. But the fact that these are externalised in the Cp helper, and not on the Element class makes things tough to extend.

In my instance with Formie, I would like to swap out a status icon indicator with a "spam" and "incomplete" icon like entries do with "drafts".

<img width="163" alt="image" src="https://github.com/craftcms/cms/assets/1221575/a5c47bcd-fd7b-4316-a1fd-5339244b6234">

So it would be _great_ to just override [`statusIndicatorHtml()`](https://github.com/craftcms/cms/blob/7ce8901f3fbd7982011b6892e8596116664d93b6/src/helpers/Cp.php#L625-L649) and add a few cases in there like there is for `draft`, but that's not really possible, as it's in the helper.

So, I suppose there's a few options:
- Add more events to the Cp helper class to modify all components of the label/chip
- Move methods, or defer to element-class-based methods for these functions, so custom element types can freely modify them

For now, the only approach is to override the chip HTML via an event:

```
 Event::on(Cp::class, Cp::EVENT_DEFINE_ELEMENT_CHIP_HTML, [Submission::class, 'defineElementChipHtml']);
```

Which is certainly fine, if somewhat roundabout - but I still need to re-construct a **lot** of functions that are private in the Cp helper class. For example, I can't call `elementLabelHtml()` or `baseElementAttributes()` which would make overriding just the status HTML pretty easy.

Finally, the `EVENT_DEFINE_ELEMENT_CHIP_HTML` event doesn't pass through the original `$config` variable, or the `$attributes` variable, which would be neat to include, to pass on to these functions like `elementLabelHtml()` and `baseElementAttributes()`.

With Craft 5 still being in alpha, maybe I'm missing something incredibly obvious, or things are still yet to be determined, so correct me!

Just to illustrate what the working solution is - there's a lot of stuff!! And I've gutted things to just suit Formie's Submissions needs.

```php
public static function defineElementChipHtml(\craft\events\DefineElementHtmlEvent $event): void
{
    $element = $event->element;

    if (!($element instanceof self)) {
        return;
    }

    $elementsService = Craft::$app->getElements();
    $user = Craft::$app->getUser()->getIdentity();
    $editable = $user && $elementsService->canView($element, $user);

    $id = sprintf('chip-%s', mt_rand());

    $attributes = array_merge($element->getHtmlAttributes($event->context), [
        'class' => ['element', 'chip', 'small'],
        'title' => $element->getUiLabel(),
        'id' => $id,
        'data' => [
            'type' => get_class($element),
            'id' => $element->id,
            'draft-id' => $element->draftId,
            'revision-id' => $element->revisionId,
            'site-id' => $element->siteId,
            'status' => $element->getStatus(),
            'label' => (string)$element,
            'url' => $element->getUrl(),
            'cp-url' => $editable ? $element->getCpEditUrl() : null,
            'level' => $element->level,
            'trashed' => $element->trashed,
            'editable' => $editable,
            'savable' => $editable && $elementsService->canSave($element),
            'duplicatable' => $editable && $elementsService->canDuplicate($element),
            'deletable' => $editable && $elementsService->canDelete($element),

            'settings' => [
                'selectable' => false,
                'context' => $event->context,
                'id' => Craft::$app->getView()->namespaceInputId($id),
                'showDraftName' => true,
                'showLabel' => true,
                'showStatus' => true,
                'showThumb' => false,
                'size' => 'small',
                'ui' => 'chip',
            ],
        ],
    ]);

    $html = Html::beginTag('div', $attributes);
    $html .= Html::beginTag('div', ['class' => 'chip-content']);

    if ($element->isIncomplete) {
        $html .= Html::tag('span', '', [
            'data' => ['icon' => 'draft'],
            'class' => 'icon',
            'role' => 'img',
            'aria' => [
                'label' => sprintf('%s %s', Craft::t('app', 'Status:'), Craft::t('formie', 'Incomplete')),
            ],
        ]);
    } else if ($element->isSpam) {
        $html .= Html::tag('span', '', [
            'data' => ['icon' => 'error'],
            'class' => 'icon',
            'role' => 'img',
            'aria' => [
                'label' => sprintf('%s %s', Craft::t('app', 'Status:'), Craft::t('formie', 'Spam')),
            ],
        ]);
    } else {
        $status = $element->getStatus();
        $statusAttributes = $element::statuses()[$status] ?? null;

        // Just to give the `statusIndicatorHtml` clean types
        if (is_string($statusAttributes)) {
            $statusAttributes = ['label' => $statusAttributes];
        }

        $html .= Html::tag('span', '', [
            'class' => array_filter([
                'status',
                $status,
                $statusAttributes['color'] ?? null,
            ]),
            'role' => 'img',
            'aria' => [
                'label' => sprintf('%s %s', Craft::t('app', 'Status:'), $statusAttributes['label'] ?? ucfirst($status)),
            ],
        ]);
    }

    $html .= Html::beginTag('div', ['class' => 'label', 'id' => $id . '-label']);
    $html .= Html::tag('a', $element->getChipLabelHtml(), ['class' => 'label-link', 'href' => $element->getCpEditUrl()]);
    $html .= Html::endTag('div') . Html::endTag('div') . Html::endTag('div');

    $event->html = $html;
}
```

**Comment by brandonkelly** [comment]
Little confused. You keep referencing element chips (which are the small instances of elements used in places like relational fields), but your screenshot and some of the description makes it sound more like you just want to add additional element status filter options?

**Reply by engram-design** [reply]
Sorry for the confusion - showing the filters was probably throwing you. What I want to achieve is this:

<img width="1118" alt="image" src="https://github.com/craftcms/cms/assets/1221575/f8992aad-5ba0-4f66-8768-cef69ac7cda4">

I'm not sure if "chip" is the right term, but I'm talking about the element's UI when used in the element index table. But it would also probably be the same for when used in the context of a relational field (of which there's a Submissions element select field).

Does that make sense? I just want to add custom icons where the status indicator is, but the helper methods that make up the label are mostly private. [Here](https://github.com/verbb/formie/blob/ffd51989ae1d3e3e14ef6a1868fe2c0b9bc4c6be/src/elements/Submission.php#L324) is how I'm currently achieving it.
